sos/report/plugins/ceph_mon.py

```diff
@@ -37,32 +37,48 @@ class CephMON(Plugin, RedHatPlugin, UbuntuPlugin):
     # but by default they are not capable of running various ceph commands in
     # this plugin - the `ceph` binary is functional directly on the host
     containers = ('ceph-(.*-)?mon.*',)
-    files = ('/var/lib/ceph/mon/', '/var/lib/ceph/*/mon*')
+    files = ('/var/lib/ceph/mon/', '/var/lib/ceph/*/mon*',
+             '/var/snap/microceph/common/data/mon/*')
     ceph_version = 0
 
     def setup(self):
 
         self.ceph_version = self.get_ceph_version()
 
-        self.add_file_tags({
-            '.*/ceph.conf': 'ceph_conf',
-            "/var/log/ceph/(.*/)?ceph-.*mon.*.log": 'ceph_mon_log'
-        })
-
-        self.add_forbidden_path([
-            "/etc/ceph/*keyring*",
-            "/var/lib/ceph/**/*keyring*",
-            # Excludes temporary ceph-osd mount location like
-            # /var/lib/ceph/tmp/mnt.XXXX from sos collection.
-            "/var/lib/ceph/**/tmp/*mnt*",
-            "/etc/ceph/*bindpass*"
-        ])
+        if not self.path_exists('/var/snap/microceph/common'):
+            self.add_file_tags({
+                '.*/ceph.conf': 'ceph_conf',
+                "/var/log/ceph/(.*/)?ceph-.*mon.*.log": 'ceph_mon_log'
+            })
+
+            self.add_forbidden_path([
+                "/etc/ceph/*keyring*",
+                "/var/lib/ceph/**/*keyring*",
+                # Excludes temporary ceph-osd mount location like
+                # /var/lib/ceph/tmp/mnt.XXXX from sos collection.
+                "/var/lib/ceph/**/tmp/*mnt*",
+                "/etc/ceph/*bindpass*"
+            ])
+
+            self.add_copy_spec([
+                "/run/ceph/**/ceph-mon*",
+                "/var/lib/ceph/**/kv_backend",
+                "/var/log/ceph/**/*ceph-mon*.log"
+            ])
 
-        self.add_copy_spec([
-            "/run/ceph/**/ceph-mon*",
-            "/var/lib/ceph/**/kv_backend",
-            "/var/log/ceph/**/*ceph-mon*.log"
-        ])
+        else:
```

Comment:
Instead of duplicating all of the copy specs, could we calculate a path prefix instead?

e.g.
```
if microceph:
    data_dir = '/var/snap/microceph/common/data'
else:
    data_dir = '/var/lib/ceph'
  
self.add_copy_spec([
    os.path.join(data_dir, 'mon/*/store.db')
])
```

something like this used to be done for containers although I think it was since mostly replaced with the logic for looking for paths inside containers instead. so most of the code that did similar seems gone now.

Reply:
I was avoiding touching any of the existing ceph code, so apart from an else i tried to not change it. I can do it this way, sure..

Reply:
The files we collect aren't the same, for eg the calamari ones are not present in the microceph env,  so for eg, in the ceph_common now, (am using the snap detection method here)

~~~
        microceph_pkg = self.policy.package_manager.pkg_by_name('microceph')
        if not microceph_pkg:
<snip>

            if not all_logs:
                self.add_copy_spec("/var/log/calamari/*.log",)
            else:
                self.add_copy_spec("/var/log/calamari",)

            self.add_copy_spec([
                "/var/log/ceph/**/ceph.log",
                "/var/log/ceph/**/ceph.audit.log*",
                "/var/log/calamari/*.log",
                "/etc/ceph/",
                "/etc/calamari/",
                "/var/lib/ceph/tmp/",
            ])

            self.add_forbidden_path([
                "/etc/ceph/*keyring*",
                "/var/lib/ceph/*keyring*",
                "/var/lib/ceph/*/*keyring*",
                "/var/lib/ceph/*/*/*keyring*",
                "/var/lib/ceph/osd",
                "/var/lib/ceph/mon",
                # Excludes temporary ceph-osd mount location like
                # /var/lib/ceph/tmp/mnt.XXXX from sos collection.
                "/var/lib/ceph/tmp/*mnt*",
                "/etc/ceph/*bindpass*"
            ])
        else:
            self.add_copy_spec([
                "/var/snap/microceph/common/logs/ceph.log",
                "/var/snap/microceph/common/logs/ceph.audit.log",
            ])
~~~

The forbidden paths are also unnecessary for microceph since we don't collect anything but those two logs. So since the microceph log collections are smaller than ceph, and that holds true for the other ceph plugins too, I've separated them out, hence the duplicate add_copy_spec() calls.

Reply:
I agree with @lathiat, if we have common file paths, then it would be easier to change one line, rather than 2 lines foe anything we are collecting. It just means less error-prone and less items to change if there are nay changes required in the future.

imo, if a file doesn't exist, then there isn't a real problem, it will be skipped by sos anyway. So, having a common set should reduce the amount of items we add

Reply:
Looking at ceph_common as an example, I've tried making the change recommended in the review to have a common PATH variable and use only one code path for the file collections, but I still feel the code is eventually cleaner and easier to follow with the separation between the two plugins with one big "if". 

The microceph stuff in ceph_common plugin as of now only collects two log files

~~~
            self.add_copy_spec([
                "/var/snap/microceph/common/logs/ceph.log",
                "/var/snap/microceph/common/logs/ceph.audit.log",
            ])
~~~

The ceph_common plugin as of now collects more stuff, so for eg consider, the only thing in common is marked in <==
~~~
        if not microceph_pkg:
            self.add_file_tags({
                '.*/ceph.conf': 'ceph_conf',
                '/var/log/ceph(.*)?/ceph.log.*': 'ceph_log',
            })

        if not all_logs:
                self.add_copy_spec("/var/log/calamari/*.log",)
            else:
                self.add_copy_spec("/var/log/calamari",)

            self.add_copy_spec([
                "/var/log/ceph/**/ceph.log", <==
                "/var/log/ceph/**/ceph.audit.log*", <==
                "/var/log/calamari/*.log",
                "/etc/ceph/",
                "/etc/calamari/",
                "/var/lib/ceph/tmp/",
            ])

            self.add_forbidden_path([
                "/etc/ceph/*keyring*",
                "/var/lib/ceph/*keyring*",
                "/var/lib/ceph/*/*keyring*",
                "/var/lib/ceph/*/*/*keyring*",
                "/var/lib/ceph/osd",
                "/var/lib/ceph/mon",
                # Excludes temporary ceph-osd mount location like
                # /var/lib/ceph/tmp/mnt.XXXX from sos collection.
                "/var/lib/ceph/tmp/*mnt*",
~~~ 
So I think it's still more reasonable to separate it out with a "if" instead of trying to have common code. It's also easier in the future to make changes to either plugin if there's a clean separation.

If I look at ceph_osd, as another eg, 
~~~
            self.add_copy_spec([
                "/var/snap/microceph/common/data/osd/*",
                "/var/snap/microceph/common/logs/*ceph-osd*.log",
            ])
~~~
It's not necessary that microceph would collect a subset of the ceph plugin, the plugins may differ in the files they collect (and they do even now) because the /var/snap/microceph/common/data/osd/* folder in microceph would collect as of now all these files,

~~~
root@demonax:/var/snap/microceph/common/data/osd/ceph-0# ls
bfm_blocks  bfm_blocks_per_key  bfm_bytes_per_block  bfm_size  block  bluefs  ceph_fsid  fsid  keyring  kv_backend  magic  mkfs_done  ready  require_osd_release  type  whoami
root@demonax:/var/snap/microceph/common/data/osd/ceph-0# 
~~~
I think if we have a common path, it would become more complex in the future maintaining this special situation we have where we are triggering one of two plugins from one plugin file. So I'd much rather have a clean separation instead of trying to run common code using variables.

Please let me know whether this makes sense.

Reply:
ack, makes sense to me